dhalli/Main.hs

```diff
@@ -0,0 +1,235 @@
+        greeter
+    )
+    emptyEnv
+
+
+data Env = Env
+  { envBindings :: StrictMap.Map Dhall.Text Binding
+  }
+
+
+emptyEnv :: Env
+emptyEnv =
+  Env
+    { envBindings = mempty
+    }
+
+
+data Binding = Binding
+  { bindingExpr :: Dhall.Expr Dhall.Src Dhall.X
+  , bindingType :: Dhall.Expr Dhall.Src Dhall.X
+  }
+
+
+envToContext :: Env -> Dhall.Context.Context ( Dhall.Expr Dhall.Src Dhall.X )
+envToContext Env{ envBindings } =
+  StrictMap.foldlWithKey'
+    ( \ctx k b -> Dhall.Context.insert k ( bindingType b ) ctx )
+    Dhall.Context.empty
+    envBindings
+
+
+parseAndLoad
+  :: ( MonadIO m, MonadState Env m )
+  => String -> m ( Dhall.Expr Dhall.Src Dhall.X )
+parseAndLoad src = do
+  parsed <-
+    case Dhall.exprFromText ( Trifecta.Columns 0 0 ) ( LazyText.pack src ) of
+      Left e ->
+        liftIO ( throwIO e )
+
+      Right a ->
+        return a
+
+  liftIO ( Dhall.load parsed )
+
+
+eval :: ( MonadIO m, MonadState Env m ) => String -> m ()
+eval src = do
+  loaded <-
+    parseAndLoad src
+
+  exprType <-
+    typeCheck loaded
+
+  expr <-
+    normalize loaded
+
+  modify
+    ( \e ->
+        e
+          { envBindings =
+              StrictMap.insert "it" ( Binding expr exprType ) ( envBindings e )
+          }
+    )
+
+  output expr
+    
+
+
+typeOf :: ( MonadIO m, MonadState Env m ) => [String] -> m ()
+typeOf [] =
+  liftIO ( putStrLn ":type requires an argument to check the type of" )
+
+
+typeOf srcs = do
+  loaded <-
+    parseAndLoad ( unwords srcs )
+
+  exprType <-
+    typeCheck loaded
+
+  exprType' <-
+    normalize exprType
+
+  output ( Expr.Annot loaded exprType' )
+
+
+
+normalize
+  :: MonadState Env m
+  => Dhall.Expr Dhall.Src Dhall.X -> m ( Dhall.Expr t Dhall.X )
+normalize e = do
+  env <-
+    get
+
+  return
+    ( Dhall.normalize
+        ( StrictMap.foldlWithKey'
+            ( \a k expr ->
+                Dhall.subst ( Dhall.V k 0 ) ( bindingExpr expr ) a
+            )
+            e
+            ( envBindings env )
+        )
+    )
+
+
+typeCheck
+  :: ( MonadIO m, MonadState Env m )
+  => Dhall.Expr Dhall.Src Dhall.X -> m ( Dhall.Expr Dhall.Src Dhall.X )
+typeCheck expr = do
+  env <-
+    get
+
+  case Dhall.typeWith ( envToContext env ) expr of
+    Left e ->
+      liftIO ( throwIO e )
+
+    Right a ->
+      return a
+
+
+addBinding :: ( MonadIO m, MonadState Env m ) => [String] -> m ()
+addBinding (k : "=" : srcs) = do 
+  let
+    varName =
+      LazyText.pack k
+  
+  loaded <-
+    parseAndLoad ( unwords srcs )
+
+  expr <-
+    normalize loaded
+
+  t <-
+    typeCheck expr
+
+  modify
+    ( \e ->
+        e
+          { envBindings =
+              StrictMap.insert
+                varName
+                ( Binding { bindingType = t, bindingExpr = expr } )
+                ( envBindings e )
+          }
+    )
+
+  output 
+    ( Expr.Annot ( Expr.Var ( Dhall.V varName 0 ) ) t )
+
+addBinding _ =
+  error ":let should be of the form `:let x = y`"
```

Comment:
You probably want to use `liftIO . fail` instead of `error`.  The issue with `error` is that it is not evaluation-safe:

```
error message `seq` () = error message
fail  message `seq` () = ()
```